src/MagnumPlugins/UfbxImporter/UfbxImporter.cpp

```diff
@@ -0,0 +1,1013 @@
+        return {};
+    }
+
+    if (l->decay == UFBX_LIGHT_DECAY_NONE) {
+        attenuation = {1.0f, 0.0f, 0.0f};
+    } else if (l->decay == UFBX_LIGHT_DECAY_LINEAR) {
+        attenuation = {0.0f, 1.0f, 0.0f};
+    } else if (l->decay == UFBX_LIGHT_DECAY_QUADRATIC) {
+        attenuation = {0.0f, 0.0f, 1.0f};
+    } else if (l->decay == UFBX_LIGHT_DECAY_CUBIC) {
+        Warning{} << "Trade::UfbxImporter::light(): cubic attenuation not supported, patching to quadratic";
+        attenuation = {0.0f, 0.0f, 1.0f};
+    } else {
+        Error() << "Trade::UfbxImporter::light(): light type" << l->type << "is not supported";
+    }
+
+
+    if((lightType == LightData::Type::Directional || lightType == LightData::Type::Ambient) && attenuation != Vector3{1.0f, 0.0f, 0.0f}) {
+        Warning{} << "Trade::UfbxImporter::light(): patching attenuation" << attenuation << "to" << Vector3{1.0f, 0.0f, 0.0f} << "for" << lightType;
+        attenuation = {1.0f, 0.0f, 0.0f};
+    }
+
+    return LightData{lightType, color, intensity, attenuation,
+        Deg{(Float)l->inner_angle}, Deg{(Float)l->outer_angle},
+        l};
+}
+
+UnsignedInt UfbxImporter::doMeshCount() const {
+    return (UnsignedInt)_state->meshChunks.size();
+}
+
+Containers::Optional<MeshData> UfbxImporter::doMesh(UnsignedInt id, UnsignedInt level) {
+    if (level != 0) return {};
+
+    MeshChunk chunk = _state->meshChunks[id];
+    ufbx_mesh *mesh = _state->scene->meshes[chunk.meshId];
+    ufbx_mesh_material mat = mesh->materials[chunk.meshMaterialIndex];
+
+    const UnsignedInt indexCount = mat.num_triangles * 3;
+
+    /* Gather all attributes. Position is there always, others are optional */
+    std::size_t attributeCount = 1;
+    std::size_t stride = sizeof(Vector3);
+
+    if(mesh->vertex_normal.exists) {
+        ++attributeCount;
+        stride += sizeof(Vector3);
+    }
+
+    if(mesh->vertex_tangent.exists) {
+        ++attributeCount;
+        stride += sizeof(Vector3);
+    }
+
+    if(mesh->vertex_bitangent.exists) {
+        ++attributeCount;
+        stride += sizeof(Vector3);
+    }
+
+    attributeCount += mesh->uv_sets.count;
+    stride += mesh->uv_sets.count * sizeof(Vector2);
+
+    attributeCount += mesh->color_sets.count;
+    stride += mesh->color_sets.count * sizeof(Color4);
+
+    Containers::Array<UnsignedInt> triangleIndices{mesh->max_face_triangles * 3};
+
+    Containers::Array<char> vertexData{NoInit, stride*indexCount};
+    size_t vertexOffset = 0;
+
+    for(UnsignedInt faceIndex : mat.face_indices) {
+        ufbx_face face = mesh->faces[faceIndex];
+        UnsignedInt numTriangles = ufbx_triangulate_face(triangleIndices.data(), triangleIndices.size(), mesh, face);
+        UnsignedInt numIndices = numTriangles * 3;
+
+        for(UnsignedInt i = 0; i < numIndices; i++) {
+            UnsignedInt ix = triangleIndices[i];
+
+            pushVector3(vertexData, vertexOffset, mesh->vertex_position[ix]);
```

Comment:
The ideal workflow here would be to form the `StridedArrayView`s (that you have below) first, and then simply assigning to them:

```cpp
Containers::StridedArrayView1D<Vector3> positions = ...;

for(...) {
    for(...) {
        positions[offset] = mesh->vertex_position[ix];
        ....
    }
}

CORRADE_INTERNAL_ASSERT(offset == indexCount);
```

There are also tools like [MeshTools::interleavedLayout()](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1MeshTools.html#a677cfe77fac62c88a02e3bb9b7afca0d) which would create an uninitialized MeshData out of a desired set of attributes and you'd then fill those, but I think the use case here is simple enough to not need to depend on the MeshTools library.

Reply:
Nice, that is a lot cleaner than writing to memory and hoping they end up in the same layout!

Reply:
Hmm as the vertex data shrinks it won't be quite as neat, but I'll try to do something.

By the way, I just followed what `AssimpImporter` and I include tangents/bitangents from only the first UV set (aka layer), but FBX (and ufbx) supports separate tangent space for each, so I could include those. Maybe a configuration option?

Reply:
Re vertex data shrinking, see [my other comment](https://github.com/mosra/magnum-plugins/pull/133#discussion_r1027297138) -- this would then work with it as well.

Separate tangent space for each UV set makes total sense, actually. Include them as additional instances of the tangents/bitangents attributes same as with texture coordinates. No option needed, if the users wouldn't want them they can filter them away.

Hmm, now I wonder how one is supposed to represent this elsewhere, same as Assimp glTF also has support for multiple texture coordinate sets but only one tangent space. What is the format supposed to do if the additional UVs lead to some completely different tangents?

Reply:
Yea it's tricky as tangent space is usually induced by UV coordinates (ie. derivatives in position for `UV.x` and `UV.y`) so having a single set of tangents for multiple UVs doesn't make much semantic sense. I guess most formats just report tangents for the first layer.

Reply:
There's actually a slight problem with the implicit mapping of UVs to tangents. There is no way to represent a mesh which only has tangents for the second UV set, I don't think this can happen in practice (again, "never seen a file that does this") so I think I'll just include tangents from UV sets up to the first one that doesn't have them, that should be 100% safe even with the parallel attributes.

Reply:
I'm aware of the implicit attribute mapping issue, and I don't have a solution yet.

Skipping tangents that have no matching UVs seems like a good idea. Printing a warning in that case might make sense though, to avoid silent data loss.